// send-crypto/build/module/handlers/TERRA/TERRAHandler.js
import BigNumber from "bignumber.js";
import { Denom, LCDClient, MsgSend, RawKey, } from "@terra-money/terra.js";
import { newPromiEvent } from "../../lib/promiEvent";
import { strip0x } from "../../lib/utils";
export var TerraNetwork;
(function (TerraNetwork) {
    TerraNetwork["Tequila"] = "tequila-0004";
    TerraNetwork["Columbus"] = "columbus-3";
})(TerraNetwork || (TerraNetwork = {}));
const toDenom = (asset) => Denom.LUNA;
export class TERRAHandler {
    constructor(privateKey, network, options = {}, sharedState) {
        this.decimals = 6;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["LUNA"].indexOf(asset.toUpperCase()) !== -1;
        this.address = (asset, _options = {}) => {
            if (!this.handlesAsset(asset)) {
                throw new Error(`Asset ${asset} not supported.`);
            }
            return this.key.accAddress;
        };
        // Balance
        this.getBalance = async (asset, options = {}) => (await this.getBalanceInSats(asset, options)).dividedBy(new BigNumber(10).exponentiatedBy(this.decimals));
        this.getBalanceInSats = async (asset, options = {}) => {
            if (!this.handlesAsset(asset)) {
                throw new Error(`Asset ${asset} not supported.`);
            }
            const balances = (await this.client.bank.balance((options && options.address) || (await this.address(asset)))).get(toDenom(asset));
            return new BigNumber(balances ? balances.amount.toFixed() : 0);
        };
        // Transfer
        this.send = (to, value, asset, options = {}) => this.sendSats(to, value.times(new BigNumber(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, value, asset, options = {}) => {
            const promiEvent = newPromiEvent();
            (async () => {
                const address = await this.address(asset);
                const send = new MsgSend(address, to, { uluna: value.toFixed() });
                const signedTx = await this.wallet.createAndSignTx({
                    msgs: [send],
                    memo: options.memo,
                });
                const result = await this.client.tx.broadcast(signedTx);
                const txHash = result.txhash;
                promiEvent.emit("transactionHash", txHash);
                promiEvent.resolve(txHash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.network =
            network === "mainnet"
                ? TerraNetwork.Columbus
                : TerraNetwork.Tequila;
        const client = sharedState.client ||
            new LCDClient({
                URL: options && options.terra ? options.terra.URL : "",
                chainID: this.network,
            });
        const key = new RawKey(Buffer.from(strip0x(privateKey), "hex"));
        const wallet = client.wallet(key);
        this.client = client;
        this.wallet = wallet;
        this.key = key;
    }
}
