// send-crypto/build/main/lib/promiEvent.js
"use strict";
// Modified from web3-core-method
Object.defineProperty(exports, "__esModule", { value: true });
exports.forwardEvents = exports.newPromiEvent = void 0;
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file PromiEvent.js
 * @author Fabian Vogelsteller <fabian@ethereum.org>, Samuel Furter <samuel@ethereum.org>
 * @date 2018
 */
// tslint:disable: no-any
const events_1 = require("events");
class InternalPromiEvent {
    /**
     * @constructor
     */
    constructor() {
        // tslint:disable-next-line: promise-must-complete
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
        this.eventEmitter = new events_1.EventEmitter();
        return new Proxy(this, {
            get: this.proxyHandler,
        });
    }
    /**
     * Proxy handler to call the promise or eventEmitter methods
     */
    proxyHandler(target, name) {
        if (name === "resolve" || name === "reject") {
            return target[name];
        }
        if (name === "then") {
            return target.promise.then.bind(target.promise);
        }
        if (name === "catch") {
            return target.promise.catch.bind(target.promise);
        }
        if (target.eventEmitter[name]) {
            return target.eventEmitter[name];
        }
    }
}
const newPromiEvent = () => new InternalPromiEvent();
exports.newPromiEvent = newPromiEvent;
const forwardEvents = (src, dest, filterFn = (_name) => true) => {
    // const forwardEmitterNewListener = (eventName: string, listener: (...args: any[]) => void) => {
    //     if (filterFn(eventName) && listener.name.indexOf("__forward_emitter_") !== 0) {
    //         console.log(`Forwarding ${eventName} Listener:`);
    //         console.log(listener);
    //         src.on(eventName, listener);
    //         src.on("transactionHash", (txHash) => { console.log(`Got transaction hash on src`); });
    //     } else {
    //         console.log("Can't forward PromiEvent event!");
    //     }
    // };
    // const forwardEmitterRemoveListener = (eventName: string, listener: (...args: any[]) => void) => {
    //     src.removeListener(eventName, listener);
    // };
    // // Listeners bound to the destination emitter should be bound to the source emitter.
    // dest.on("newListener", forwardEmitterNewListener);
    // // When a listener is removed from the destination emitter, remove it from the source emitter
    // // (otherwise it will continue to be called).
    // dest.on("removeListener", forwardEmitterRemoveListener);
    // Until the above is fixed, we manually forward each event name:
    src.on("transactionHash", (eventReceipt) => {
        dest.emit("transactionHash", eventReceipt);
    });
    src.on("receipt", (eventReceipt) => {
        dest.emit("receipt", eventReceipt);
    });
    src.on("confirmation", (confNumber, eventReceipt) => {
        dest.emit("confirmation", confNumber, eventReceipt);
    });
    src.on("error", (error) => {
        dest.emit("error", error);
    });
};
exports.forwardEvents = forwardEvents;
