// send-crypto/build/module/handlers/FIL/FILHandler.js
import BigNumber from "bignumber.js";
import { Message } from "@glif/filecoin-message";
import Filecoin from "@glif/filecoin-wallet-provider";
import { Network as FilNetwork } from "@glif/filecoin-address";
import { newPromiEvent } from "../../lib/promiEvent";
import { sleep } from "../../lib/retry";
import { SingleKeyProvider } from "./provider";
export class FILHandler {
    constructor(privateKey, network, options = {}, sharedState) {
        this.decimals = 18;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["FIL", "FILECOIN"].indexOf(asset.toUpperCase()) !== -1;
        this.address = async (asset, _options = {}) => {
            if (!this.handlesAsset(asset)) {
                throw new Error(`Asset ${asset} not supported.`);
            }
            const accounts = await this.sharedState.filecoin.wallet.getAccounts(0, 1, this.network);
            return accounts[0];
        };
        // Balance
        this.getBalance = async (asset, options = {}) => new BigNumber(await this.sharedState.filecoin.getBalance((options && options.address) || (await this.address(asset))));
        this.getBalanceInSats = async (asset, options = {}) => (await this.getBalance(asset, options)).multipliedBy(new BigNumber(10).exponentiatedBy(this.decimals));
        // Transfer
        this.send = (to, value, asset, options = {}) => this.sendSats(to, value.times(new BigNumber(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, valueIn, asset, options = {}) => {
            const promiEvent = newPromiEvent();
            (async () => {
                const address = await this.address(asset);
                const nonce = await this.sharedState.filecoin.getNonce(address);
                let value = valueIn;
                let message = new Message({
                    to,
                    from: address,
                    value: valueIn,
                    method: 0,
                    gasFeeCap: 0,
                    gasLimit: 0,
                    gasPremium: 0,
                    nonce,
                    params: undefined,
                    ...options,
                });
                message = await this.sharedState.filecoin.gasEstimateMessageGas(message.toLotusType());
                if (options.subtractFee) {
                    const { maxFee: fee, } = await this.sharedState.filecoin.gasEstimateMaxFee(message.toLotusType());
                    if (fee.gt(value)) {
                        throw new Error(`Unable to include fee in value, fee exceeds value (${fee.toFixed()} > ${value.toFixed()})`);
                    }
                    value = value.minus(fee);
                }
                const signed = await this.sharedState.filecoin.wallet.sign(address, message.toLotusType());
                const tx = await this.sharedState.filecoin.sendMessage(signed.Message, signed.Signature.Data);
                const txHash = tx["/"];
                promiEvent.emit("transactionHash", txHash);
                while (true) {
                    const result = await this.sharedState.filecoin.jsonRpcEngine.request("StateSearchMsg", tx);
                    /*
                    {
                        Message: {
                            '/': 'bafy2bzacedtcalpfb5dj3de6jb5lfncqtxw4kia4xotpg3on3p7gcesiwlube'
                        },
                        Receipt: { ExitCode: 0, Return: null, GasUsed: 433268 },
                        ReturnDec: null,
                        TipSet: [
                            {
                            '/': 'bafy2bzacedsqymqnjl2oir7q2rgbwyqscihvebkvr7hfzdq5b2tvcct3cwczy'
                            },
                            {
                            '/': 'bafy2bzacednrjqbsyooob3f76wwct4iupauehvq6wnzrbf6uc4dskptwu4er6'
                            },
                            {
                            '/': 'bafy2bzaceany2in5cadbyf27mecnmzc6ustqfewd5dobxquidzbn3wtnjd52y'
                            }
                        ],
                        Height: 53025
                    }
                    */
                    if (result) {
                        promiEvent.emit("confirmation", 1);
                        break;
                    }
                    await sleep(10 * 1000);
                }
                promiEvent.resolve(txHash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        const filecoin = new Filecoin(SingleKeyProvider(privateKey), {
            apiAddress: options && options.apiAddress,
            token: options && options.token,
        });
        this.network =
            network === "mainnet" ? FilNetwork.MAIN : FilNetwork.TEST;
        sharedState.filecoin = filecoin;
        this.sharedState = sharedState;
    }
}
