// send-crypto/build/main/handlers/ETH/ethUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTransactionConfig = exports.getNetwork = exports.Network = exports.getEndpoint = exports.getWeb3 = void 0;
const web3_1 = __importDefault(require("web3"));
const getWeb3 = (privateKey, endpoint) => {
    // const provider = new HDWalletProvider(privateKey, endpoint);
    const web3 = new web3_1.default(endpoint);
    const account = web3.eth.accounts.privateKeyToAccount("0x" + privateKey);
    web3.eth.accounts.wallet.add(account);
    // tslint:disable-next-line: no-object-mutation
    web3.eth.defaultAccount = account.address;
    // return new Web3(provider as any);
    return [web3, account.address];
};
exports.getWeb3 = getWeb3;
// Free tier - only used as a fallback.
const defaultInfuraKey = "3b7a6c29f9c048d688a848899888aa96";
const getEndpoint = (network, ethereumNode, infuraKey) => {
    return ethereumNode
        ? ethereumNode
        : `https://${network}.infura.io/v3/${infuraKey || defaultInfuraKey}`;
};
exports.getEndpoint = getEndpoint;
var Network;
(function (Network) {
    Network["Mainnet"] = "mainnet";
    Network["Ropsten"] = "ropsten";
    Network["Kovan"] = "kovan";
    Network["Rinkeby"] = "rinkeby";
    Network["G\u00F6rli"] = "goerli";
})(Network = exports.Network || (exports.Network = {}));
const getNetwork = (network) => {
    switch (network.toLowerCase()) {
        case "mainnet":
        case "main":
            return Network.Mainnet;
        case "kovan":
            return Network.Kovan;
        case "rinkeby":
            return Network.Rinkeby;
        case "görli":
        case "goerli":
        case "gorli":
            return Network.Görli;
        case "ropsten":
        case "testnet":
        default:
            return Network.Ropsten;
    }
};
exports.getNetwork = getNetwork;
const getTransactionConfig = (options) => {
    const txConfig = {};
    // tslint:disable: no-object-mutation
    if (options.from) {
        txConfig.from = options.from;
    }
    if (options.from) {
        txConfig.from = options.from;
    }
    // if (options.to) { txConfig.to = options.to; }
    if (options.gasPrice) {
        txConfig.gasPrice = options.gasPrice;
    }
    if (options.gas) {
        txConfig.gas = options.gas;
    }
    if (options.value) {
        txConfig.value = options.value;
    }
    if (options.data) {
        txConfig.data = options.data;
    }
    if (options.nonce) {
        txConfig.nonce = options.nonce;
    }
    // if (options.chainId) { txConfig.chainId = options.chainId; }
    // if (options.common) { txConfig.common = options.common; }
    // if (options.chain) { txConfig.chain = options.chain; }
    // if (options.hardfork) { txConfig.hardfork = options.hardfork; }
    return txConfig;
};
exports.getTransactionConfig = getTransactionConfig;
